pkg/resourcemapping/operator_resource_mapping.go

```diff
@@ -0,0 +1,340 @@
+//           "destPath": ".spec.{{.componentName}}.resources"
+//         },
+//        ]
+//    }
+//  }
+//}
+func (ormClient *ORMClient) CacheORMSpecMap() int {
+	ormCRs, err := ormClient.getORMCRList()
+	if err != nil {
+		glog.Warningf("No OperatorResourceMapping CR discovered: %v. Create operator-resource-mapping CR to control "+
+			"Operator managed resources.", err)
+		return 0
+	}
+	ormClient.cacheLock.Lock()
+	defer ormClient.cacheLock.Unlock()
+	// Clear existing cached operatorResourceSpecMap data
+	ormClient.operatorResourceSpecMap = make(map[string]*ORMSpec)
+	for _, ormCR := range ormCRs {
+		// An orm CR name always refers to a Operator manged CRD name
+		crdName := ormCR.GetName()
+		namespace := ormCR.GetNamespace()
+		// Get Operator CRs of the CRD in the same namespace of ORM CR.
+		operatorCRs, gvRes, err := ormClient.getOperatorCRsFromCRD(crdName, namespace)
+		if err != nil {
+			glog.Errorf("Failed to get Operator managed CRs from CRD %s, %v", crdName, err)
+			continue
+		}
+		ormTemplateMap, err := ormClient.populateORMTemplateMap(ormCR)
+		if err != nil {
+			glog.Errorf("Failed to populate ormTemplateMap for CRD %s: %v", crdName, err)
+			continue
+		}
+		for _, operatorCR := range operatorCRs {
+			crUID := string(operatorCR.GetUID())
+			operatorResourceSpec, exists := ormClient.operatorResourceSpecMap[crUID]
+			if exists {
+				glog.Errorf("Skip the duplicate CR UID '%s' for CRD %s", crUID, crdName)
+				continue
+			} else {
+				operatorResourceSpec = &ORMSpec{
+					operatorGVResource: gvRes,
+					ormTemplateMap:     ormTemplateMap,
+				}
+				ormClient.operatorResourceSpecMap[crUID] = operatorResourceSpec
+			}
+		}
+	}
+	return len(ormClient.operatorResourceSpecMap)
+}
+
+func (ormClient *ORMClient) getORMCRList() ([]unstructured.Unstructured, error) {
+	groupVersionRes := schema.GroupVersionResource{
+		Group:    ormGroup,
+		Version:  ormVersion,
+		Resource: ormResource,
+	}
+	ormCRs, err := ormClient.dynClient.Resource(groupVersionRes).List(metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get OperatorResourceMapping CRs: %v", err)
+	}
+	return ormCRs.Items, nil
+}
+
+func (ormClient *ORMClient) getOperatorCRsFromCRD(crdName, namespace string) ([]unstructured.Unstructured, schema.GroupVersionResource, error) {
+	crds := ormClient.apiExtClient.CustomResourceDefinitions()
+	crd, err := crds.Get(crdName, metav1.GetOptions{})
+	if err != nil {
+		return nil, schema.GroupVersionResource{}, err
+	}
+	groupVersionRes := schema.GroupVersionResource{
+		Group:    crd.Spec.Group,
+		Version:  crd.Spec.Versions[0].Name,
+		Resource: crd.Spec.Names.Plural,
+	}
+	crs, err := ormClient.dynClient.Resource(groupVersionRes).Namespace(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return nil, groupVersionRes, err
+	}
+	return crs.Items, groupVersionRes, nil
+}
+
+func (ormClient *ORMClient) populateORMTemplateMap(ormCR unstructured.Unstructured) (map[string]ORMTemplate, error) {
```

Comment:
Its allright that you have manually curated the crd definition and read the definition into a code structure via this function for now. A more sustainable mechanism would be to use a project like [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) and bootstrap the crd type definition (ORM in our case) as golang structs.  
There are many benefits we would have with that. We can have typed clients (typed clients have all the type checking benefits) to deal with our API spec (ORM in our case) or directly use encode/decode helpers from unstructured libraries if we stick to dynamic client. Additionally any future changes could mean simply updating the golang struct definition and autogenerate the CRD spec, which can otherwise be daunting to manage manually.
Also, I remember that I did mention this to you earlier, but I must apologise to not have stressed its usage enough. I now feel that it should have made your work easier and certainly makes future maintainance easier.

Reply:
Thanks @irfanurrehman ! Yes, I remember you mentioned the "kubebuilder". And sorry I didn't fully investigate and understand the usage. When taking an initial glance at that project, I thought it might complicate the process since it provides a whole set of customization to create our own controller and resources, where most of stuff we don't need. After playing with the "Hello Word". I agree that it'll actually be more sustainable to use this `kubebuilder`. 

If I understand correctly, we would need a new project using `kubebuilder` to maintain the ORM type definition and provide go struct with ORM specs. As you said, we can not only make use of the typed clients (or directly use `unstructured` libraries), but also autogenerate the CRD spec from the golang struct. Furthermore the repo of this project could be the place where we maintain the ORM CRD and CR sample/example.

If it's ok, I'll take this as a further improvement to current logic.

Reply:
As discussed offline, lets keep this item in the backlog and revisit this when we have real customer requirements on this and are sure that we will need to maintain this for long.